### PR TITLE
fix: set x-deployment-id to every middleware prefetch request

### PR DESCRIPTION
**Pull request by abvthecity**
### What?

`x-deployment-id` is missing on prefetch requests

### Why?

To handle version skew on the pages router. When version skew is turned on, `x-deployment-id` should be used on prefetch otherwise the server will return 404. But with middleware enabled, the 404 does not appear to hard-refresh the page.

### How?

next.js seems to be relying on the 404 to perform a _hard_ navigation and auto-refresh the client. But that doesn't appear to happen when middleware is in use (separate issue, not addressed in this PR), which is causing the client to navigate to 404 whenever we deploy a new version of the site to vercel.

instead of sending the client to 404, passing the `x-deployment-id` helps guarantee that next-data requests respect version skew protection.

### CODE REVIEW

**Comment by ijjk** [comment]
Allow CI Workflow Run

- [ ] approve CI run for commit: 41f5d9cf0f1329fc6c2264d09d69998d9b497a7e

Note: this should only be enabled once the PR is ready to go and can only be enabled by a maintainer

File: packages/next/src/shared/lib/router/router.ts [hunk 1]
```diff
@@ -497,7 +497,12 @@ function fetchNextData({
       headers: Object.assign(
         {} as HeadersInit,
         isPrefetch ? { purpose: 'prefetch' } : {},
-        isPrefetch && hasMiddleware ? { 'x-middleware-prefetch': '1' } : {}
+        isPrefetch && hasMiddleware
+          ? {
+              'x-middleware-prefetch': '1',
+              'x-deployment-id': process.env.NEXT_DEPLOYMENT_ID,
```
Comment (ztanner) [inline packages/next/src/shared/lib/router/router.ts:503]: I think we should just ensure this header is always passed for requests, even if they're not prefetch / middleware is present.

Comment (ijjk) [reply packages/next/src/shared/lib/router/router.ts:503]: Just bumping this, if you're able to make this change we could re-review!

Comment (abvthecity) [reply packages/next/src/shared/lib/router/router.ts]: Thanks for bumping! Updated the PR 👍

File: packages/next/src/shared/lib/router/router.ts [hunk 2]
```diff
@@ -498,7 +498,8 @@ function fetchNextData({
       headers: Object.assign(
         {} as HeadersInit,
         isPrefetch ? { purpose: 'prefetch' } : {},
-        isPrefetch && hasMiddleware ? { 'x-middleware-prefetch': '1' } : {}
+        isPrefetch && hasMiddleware ? { 'x-middleware-prefetch': '1' } : {},
+        { 'x-deployment-id': process.env.NEXT_DEPLOYMENT_ID }
```
Comment (ijjk) [inline packages/next/src/shared/lib/router/router.ts:502]:
```suggestion
        process.env.NEXT_DEPLOYMENT_ID ? { 'x-deployment-id': process.env.NEXT_DEPLOYMENT_ID } : {}
```
We should only set the header if it has a value.

**Review by ijjk — COMMENTED** [review]
Looks like lint fixes just need applied can run `pnpm lint-fix` for this. Seems we can't as it's a PR from an organization which doesn't allow us to push.

**Comment by ijjk** [comment]
## Failing test suites 
Commit: c2a77f16221f167009f2a2d2db15eff63d4ec9ce

`pnpm test-start test/production/deployment-id-handling/deployment-id-handling.test.ts`
- deployment-id-handling enabled with CUSTOM_DEPLOYMENT_ID > should contain deployment id in prefetch request
- deployment-id-handling enabled with NEXT_DEPLOYMENT_ID > should contain deployment id in prefetch request

<details>
<summary>Expand output</summary>

  ● deployment-id-handling enabled with NEXT_DEPLOYMENT_ID › should contain deployment id in prefetch request

    expect(received).toBeGreaterThan(expected)

    Expected: > 0
    Received:   0

      100 |       await browser.elementById('edge-link').click()
      101 |
    > 102 |       expect(dataHeaders.length).toBeGreaterThan(0)
          |                                  ^
      103 |
      104 |       expect(
      105 |         dataHeaders.every(

      at Object.toBeGreaterThan (production/deployment-id-handling/deployment-id-handling.test.ts:102:34)

  ● deployment-id-handling enabled with CUSTOM_DEPLOYMENT_ID › should contain deployment id in prefetch request

    expect(received).toBeGreaterThan(expected)

    Expected: > 0
    Received:   0

      100 |       await browser.elementById('edge-link').click()
      101 |
    > 102 |       expect(dataHeaders.length).toBeGreaterThan(0)
          |                                  ^
      103 |
      104 |       expect(
      105 |         dataHeaders.every(

      at Object.toBeGreaterThan (production/deployment-id-handling/deployment-id-handling.test.ts:102:34)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ijjk** [comment]
# Stats from current PR

<details open>
<summary><strong>Default Build</strong></summary>

<br/>

<details>
<summary><strong>General</strong></summary>

|  | vercel/next.js canary  | fern-api/next.js ajiang/fix-x-deployment-id-prefetch | Change |
| - | - | - | - |
| buildDuration | 17.7s | 18.8s | ⚠️ +1.2s |
| buildDurationCached | 15.2s | 12.6s | N/A |
| nodeModulesSize | 416 MB | 410 MB | N/A |
| nextStartRea..uration (ms) | 469ms | 473ms | N/A |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | fern-api/next.js ajiang/fix-x-deployment-id-prefetch | Change |
| - | - | - | - |
| 1187-HASH.js gzip | 52.4 kB | 51.1 kB | N/A |
| 8276.HASH.js gzip | 169 B | 168 B | N/A |
| 8377-HASH.js gzip | 5.36 kB | 5.36 kB | N/A |
| bccd1874-HASH.js gzip | 53 kB | 53 kB | N/A |
| framework-HASH.js gzip | 57.5 kB | 57.5 kB | N/A |
| main-app-HASH.js gzip | 232 B | 235 B | N/A |
| main-HASH.js gzip | 34.1 kB | 34 kB | N/A |
| webpack-HASH.js gzip | 1.71 kB | 1.71 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | fern-api/next.js ajiang/fix-x-deployment-id-prefetch | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 39.4 kB | 39.4 kB | ✓ |
| Overall change | 39.4 kB | 39.4 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | fern-api/next.js ajiang/fix-x-deployment-id-prefetch | Change |
| - | - | - | - |
| _app-HASH.js gzip | 193 B | 193 B | ✓ |
| _error-HASH.js gzip | 193 B | 193 B | ✓ |
| amp-HASH.js gzip | 512 B | 510 B | N/A |
| css-HASH.js gzip | 343 B | 342 B | N/A |
| dynamic-HASH.js gzip | 1.84 kB | 1.84 kB | ✓ |
| edge-ssr-HASH.js gzip | 265 B | 265 B | ✓ |
| head-HASH.js gzip | 363 B | 362 B | N/A |
| hooks-HASH.js gzip | 393 B | 392 B | N/A |
| image-HASH.js gzip | 4.49 kB | 4.49 kB | N/A |
| index-HASH.js gzip | 268 B | 268 B | ✓ |
| link-HASH.js gzip | 2.35 kB | 2.34 kB | N/A |
| routerDirect..HASH.js gzip | 328 B | 328 B | ✓ |
| script-HASH.js gzip | 397 B | 397 B | ✓ |
| withRouter-HASH.js gzip | 323 B | 326 B | N/A |
| 1afbb74e6ecf..834.css gzip | 106 B | 106 B | ✓ |
| Overall change | 3.59 kB | 3.59 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | fern-api/next.js ajiang/fix-x-deployment-id-prefetch | Change |
| - | - | - | - |
| _buildManifest.js gzip | 749 B | 746 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | fern-api/next.js ajiang/fix-x-deployment-id-prefetch | Change |
| - | - | - | - |
| index.html gzip | 522 B | 522 B | ✓ |
| link.html gzip | 538 B | 536 B | N/A |
| withRouter.html gzip | 519 B | 519 B | ✓ |
| Overall change | 1.04 kB | 1.04 kB | ✓ |

</details>

<details>
<summary><strong>Edge SSR bundle Size</strong></summary>

|  | vercel/next.js canary  | fern-api/next.js ajiang/fix-x-deployment-id-prefetch | Change |
| - | - | - | - |
| edge-ssr.js gzip | 129 kB | 128 kB | N/A |
| page.js gzip | 205 kB | 204 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Middleware size</strong></summary>

|  | vercel/next.js canary  | fern-api/next.js ajiang/fix-x-deployment-id-prefetch | Change |
| - | - | - | - |
| middleware-b..fest.js gzip | 672 B | 668 B | N/A |
| middleware-r..fest.js gzip | 155 B | 156 B | N/A |
| middleware.js gzip | 31.3 kB | 31.2 kB | N/A |
| edge-runtime..pack.js gzip | 844 B | 844 B | ✓ |
| Overall change | 844 B | 844 B | ✓ |

</details>

<details>
<summary><strong>Next Runtimes</strong></summary>

|  | vercel/next.js canary  | fern-api/next.js ajiang/fix-x-deployment-id-prefetch | Change |
| - | - | - | - |
| 274-experime...dev.js gzip | 322 B | 322 B | ✓ |
| 274.runtime.dev.js gzip | 314 B | 314 B | ✓ |
| app-page-exp...dev.js gzip | 356 kB | 323 kB | N/A |
| app-page-exp..prod.js gzip | 129 kB | 127 kB | N/A |
| app-page-tur..prod.js gzip | 142 kB | 140 kB | N/A |
| app-page-tur..prod.js gzip | 137 kB | 135 kB | N/A |
| app-page.run...dev.js gzip | 346 kB | 314 kB | N/A |
| app-page.run..prod.js gzip | 125 kB | 123 kB | N/A |
| app-route-ex...dev.js gzip | 37.5 kB | 37.4 kB | N/A |
| app-route-ex..prod.js gzip | 25.5 kB | 25.5 kB | N/A |
| app-route-tu..prod.js gzip | 25.5 kB | 25.5 kB | N/A |
| app-route-tu..prod.js gzip | 25.3 kB | 25.3 kB | N/A |
| app-route.ru...dev.js gzip | 39.1 kB | 39 kB | N/A |
| app-route.ru..prod.js gzip | 25.3 kB | 25.3 kB | N/A |
| pages-api-tu..prod.js gzip | 9.69 kB | 9.69 kB | ✓ |
| pages-api.ru...dev.js gzip | 11.6 kB | 11.6 kB | ✓ |
| pages-api.ru..prod.js gzip | 9.68 kB | 9.68 kB | ✓ |
| pages-turbo...prod.js gzip | 21.7 kB | 21.7 kB | N/A |
| pages.runtim...dev.js gzip | 27.5 kB | 27.4 kB | N/A |
| pages.runtim..prod.js gzip | 21.7 kB | 21.7 kB | N/A |
| server.runti..prod.js gzip | 916 kB | 916 kB | N/A |
| Overall change | 31.6 kB | 31.6 kB | ✓ |

</details>

<details>
<summary><strong>build cache</strong></summary>

|  | vercel/next.js canary  | fern-api/next.js ajiang/fix-x-deployment-id-prefetch | Change |
| - | - | - | - |
| 0.pack gzip | 2.08 MB | 2.06 MB | N/A |
| index.pack gzip | 73.9 kB | 72.8 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Diff details</strong></summary>

<details>
<summary>Diff for <strong>middleware.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>edge-ssr.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>1187-HASH.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>main-HASH.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-page-exp..ntime.dev.js</strong></summary>

```diff
failed to diff
```
</details>
<details>
<summary>Diff for <strong>app-page-exp..time.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-page-tur..time.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-page-tur..time.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-page.runtime.dev.js</strong></summary>

```diff
failed to diff
```
</details>
<details>
<summary>Diff for <strong>app-page.runtime.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-route-ex..ntime.dev.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-route-ex..time.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-route-tu..time.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-route-tu..time.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-route.runtime.dev.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-route.ru..time.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>pages-turbo...time.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>pages.runtime.dev.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>pages.runtime.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>server.runtime.prod.js</strong></summary>

Diff too large to display
</details>

</details>

</details>
Commit: c2a77f16221f167009f2a2d2db15eff63d4ec9ce

**Review by ijjk — APPROVED** [review]

**Pull request merged**